java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -87,95 +180,154 @@ class EntityHandleManager(
      */
     suspend fun <T : Entity> createCollectionHandle(
         mode: HandleMode,
-        name: String,
+        baseName: String,
         entitySpec: EntitySpec<T>,
         storageKey: StorageKey,
         idGenerator: Id.Generator = Id.Generator.newSession()
-    ) = handleManager.rawEntityCollectionHandle(
-            storageKey,
+    ): BaseHandleAdapter {
+        val name = idGenerator.newChildId(
+            idGenerator.newChildId(arcId.toArcId(), hostId),
+            baseName
+        ).toString()
+
+        val entityPrep = EntityPrep<T>(
+            name,
+            idGenerator,
             entitySpec.SCHEMA,
-            name = idGenerator.newChildId(
-                idGenerator.newChildId(arcId.toArcId(), hostId),
-                name
-            ).toString()
-        ).let {
-            when (mode) {
-                HandleMode.Read -> ReadCollectionHandleAdapter(entitySpec, it)
-                HandleMode.Write -> WriteCollectionHandleAdapter<T>(it, idGenerator)
-                HandleMode.ReadWrite ->
-                    ReadWriteCollectionHandleAdapter(entitySpec, it, idGenerator)
-            }
+            Ttl.Infinite,
+            time
+        )
+
+        val store = stores.get(
+            CollectionStoreOptions<RawEntity>(
+                storageKey = storageKey,
+                type = CollectionType(EntityType(entitySpec.SCHEMA)),
+                mode = ReferenceMode
+            )
+        ).activate()
+
+        val storageProxy = collectionStorageProxies.getOrPut(storageKey) {
+            CollectionProxy(store, CrdtSet())
         }
+
+        val dereferencer: Dereferencer<RawEntity>? = RawEntityDereferencer(
+            entitySpec.SCHEMA,
+            stores,
+            activationFactory
+        )
+
+        return when (mode) {
+            HandleMode.Read ->
+                ReadCollectionHandleAdapter(
+                    name = name,
+                    entitySpec = entitySpec,
+                    storageProxy = storageProxy,
+                    dereferencer = dereferencer
+                )
+            HandleMode.Write -> WriteCollectionHandleAdapter<T>(
+                name = name,
+                storageProxy = storageProxy, entityPrep = entityPrep
+            )
+            HandleMode.ReadWrite ->
+                ReadWriteCollectionHandleAdapter(
+                    name = name,
+                    entitySpec = entitySpec,
+                    storageProxy = storageProxy,
+                    entityPrep = entityPrep,
+                    dereferencer = dereferencer
+                )
+        }
+    }
```

Comment:
I think it would be sane to split this into multiple files, eg one for everything up to here and another one for the concrete classes (also fine to do in follow up)

Reply:
Yes, this file definitely needs some re-organization. I didn't want to tackle that in this PR in the name of review sanity.